### PR TITLE
Fixes #225, Adds Startups on the map

### LINKED ISSUE #225: Integrate Angel.co and show startups and investors on a map
All details for the API are here: https://angel.co/api

### PR DESCRIPTION
**Pull request by prasht63**
- [x] Add button to add apps
- [x] Add Angel.co as an app.
- [x] Use oAuth to get token from Angel.co
- [x] Show Startups on map
- [x] Show details about the startup
- [ ] Store App Details for the user
- [ ] Allow location filter to results
Dont merge workpending

### CODE REVIEW

**Comment by prasht63** [comment]
@Orbiter I am trying to save the activated apps for a user in Loklak_Server. When I try something like
```
http://localhost:9000/api/account.json?action=update&data={%22screen_name%22:%22prasht63%22,%22apps%22:{%22dataconnect%22:{%22type%22:%22angel%22}}}
```
I get this error
```
HTTP ERROR 500

Problem accessing /api/account.json. Reason:

    Server Error
Caused by:

org.elasticsearch.index.mapper.MapperParsingException: failed to parse [apps]
	at org.elasticsearch.index.mapper.core.AbstractFieldMapper.parse(AbstractFieldMapper.java:411)
	at org.elasticsearch.index.mapper.object.ObjectMapper.serializeObject(ObjectMapper.java:554)
	at org.elasticsearch.index.mapper.object.ObjectMapper.parse(ObjectMapper.java:487)
	at org.elasticsearch.index.mapper.DocumentMapper.parse(DocumentMapper.java:544)
	at org.elasticsearch.index.mapper.DocumentMapper.parse(DocumentMapper.java:493)
	at org.elasticsearch.index.shard.IndexShard.prepareIndex(IndexShard.java:493)
	at org.elasticsearch.action.index.TransportIndexAction.shardOperationOnPrimary(TransportIndexAction.java:192)
	at org.elasticsearch.action.support.replication.TransportShardReplicationOperationAction$PrimaryPhase.performOnPrimary(TransportShardReplicationOperationAction.java:574)
	at org.elasticsearch.action.support.replication.TransportShardReplicationOperationAction$PrimaryPhase$1.doRun(TransportShardReplicationOperationAction.java:440)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:36)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.elasticsearch.ElasticsearchIllegalArgumentException: unknown property [wall]
	at org.elasticsearch.index.mapper.core.StringFieldMapper.parseCreateFieldForString(StringFieldMapper.java:331)
	at org.elasticsearch.index.mapper.core.StringFieldMapper.parseCreateField(StringFieldMapper.java:277)
	at org.elasticsearch.index.mapper.core.AbstractFieldMapper.parse(AbstractFieldMapper.java:401)
	... 12 more
Caused by:

org.elasticsearch.ElasticsearchIllegalArgumentException: unknown property [wall]
	at org.elasticsearch.index.mapper.core.StringFieldMapper.parseCreateFieldForString(StringFieldMapper.java:331)
	at org.elasticsearch.index.mapper.core.StringFieldMapper.parseCreateField(StringFieldMapper.java:277)
	at org.elasticsearch.index.mapper.core.AbstractFieldMapper.parse(AbstractFieldMapper.java:401)
	at org.elasticsearch.index.mapper.object.ObjectMapper.serializeObject(ObjectMapper.java:554)
	at org.elasticsearch.index.mapper.object.ObjectMapper.parse(ObjectMapper.java:487)
	at org.elasticsearch.index.mapper.DocumentMapper.parse(DocumentMapper.java:544)
	at org.elasticsearch.index.mapper.DocumentMapper.parse(DocumentMapper.java:493)
	at org.elasticsearch.index.shard.IndexShard.prepareIndex(IndexShard.java:493)
	at org.elasticsearch.action.index.TransportIndexAction.shardOperationOnPrimary(TransportIndexAction.java:192)
	at org.elasticsearch.action.support.replication.TransportShardReplicationOperationAction$PrimaryPhase.performOnPrimary(TransportShardReplicationOperationAction.java:574)
	at org.elasticsearch.action.support.replication.TransportShardReplicationOperationAction$PrimaryPhase$1.doRun(TransportShardReplicationOperationAction.java:440)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:36)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Powered by Jetty://
```

**Comment by aneeshd16** [comment]
@prasht63 you can use the REST API which has already been created in oauth-proxy. The endpoints are 
LIST: GET  `<oauth-proxy-url>/screen_name/app_name`
READ: GET  `<oauth-proxy-url>/screen_name/app_name/id`
CREATE: POST to `<oauth-proxy-url>/screen_name/app_name`. Response is `{ id : <randomly generated uid> }`
DELETE: DELETE to  `<oauth-proxy-url>/screen_name/app_name/id`
UPDATE: PUT to  `<oauth-proxy-url>/screen_name/app_name/id`

On the angular side, you can use ng-resource to accelerate development. I do not know if this is exactly what you need, but its a suggestion.

**Comment by mariobehling** [comment]
@prasht63 What is the status of this commit?

**Comment by prasht63** [comment]
@mariobehling This is on hold until @Orbiter comes.I am not able to store app access_token in the server. @aneeshd16 It would be really nice if you could give me an example the way you do it, or may be reference me the code where you have used it.The functionality works, its just that I need to store access_token so that everytime the user doesnot need to sign-in.

File: app/js/controllers/angel.js [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+  });
+  $scope.curr_page=0;
+  $scope.lastpage;
+ 
+
+    var url="https://api.angel.co/1/startups?filter=raising&access_token=8cd27d851f1e4b69e27a29cc9606822c118586d4c8b168b5&callback=JSON_CALLBACK";
+    console.log(url);
+    $http.jsonp(url).
+    then(function(response) {
+      $scope.curr_page=response.data.page;
```
Comment (sudheesh001) [inline app/js/controllers/angel.js]: Codacy Issue found: [Identifier 'curr_page' is not in camel case.](https://www.codacy.com/app/sudheesh1995/loklak_webclient/file/2550181458/issues/source?bid=3288768&fileBranchId=3288808#l80)

File: app/js/controllers/angel.js [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+                
+            $http.jsonp('http://localhost:9000/api/geocode.json?callback=JSON_CALLBACK&minified=true', {params : { data : locarray } })
+            .success(function(data, status, headers, config) {
+              
+              var startupsMarker = {
+                "type": "FeatureCollection",
+                "features": []
+            };
+                //console.log( followers.propic[i]);
+                $scope.startups_status=startupslocation.length;
```
Comment (sudheesh001) [inline app/js/controllers/angel.js]: Codacy Issue found: [Identifier 'startups_status' is not in camel case.](https://www.codacy.com/app/sudheesh1995/loklak_webclient/file/2550181458/issues/source?bid=3288768&fileBranchId=3288808#l163)

File: app/js/controllers/angel.js [hunk 3]
```diff
@@ -0,0 +1,253 @@
+'use strict';
+/* global angular, L */
+/* jshint unused:false */
+
+var controllersModule = require('./_index');
```
Comment (sudheesh001) [inline app/js/controllers/angel.js]: Codacy Issue found: [Strings must use doublequote.](https://www.codacy.com/app/sudheesh1995/loklak_webclient/file/2550181458/issues/source?bid=3288768&fileBranchId=3288808#l5)

File: app/js/controllers/angel.js [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+
+                        },
+                        "id": "123"
+
+                    };
+                    startupsMarker.features.push(pointObject);
+
+                    }
+                }
+                   add_marker(startupsMarker);
```
Comment (sudheesh001) [inline app/js/controllers/angel.js]: Codacy Issue found: ["add_marker" was used before it was defined](https://www.codacy.com/app/sudheesh1995/loklak_webclient/file/2550181458/issues/source?bid=3288768&fileBranchId=3288808#l197)

File: app/js/controllers/angel.js [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+                        // called asynchronously if an error occurs
+                        // or server returns response with an error status.
+            });
+               
+        }
+
+  }
+
+
+  function add_marker(result) {
```
Comment (sudheesh001) [inline app/js/controllers/angel.js]: Codacy Issue found: [Identifier 'add_marker' is not in camel case.](https://www.codacy.com/app/sudheesh1995/loklak_webclient/file/2550181458/issues/source?bid=3288768&fileBranchId=3288808#l215)

File: app/js/controllers/angel.js [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+    // called asynchronously if an error occurs
+    // or server returns response with an error status.
+  });
+
+
+    }
+   
+    
+    function Geocode()
+            { console.log("called");
```
Comment (sudheesh001) [inline app/js/controllers/angel.js]: Codacy Issue found: [Unexpected console statement.](https://www.codacy.com/app/sudheesh1995/loklak_webclient/file/2550181458/issues/source?bid=3288768&fileBranchId=3288808#l147)

File: app/js/controllers/angel.js [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+
+                    }
+                }
+                   add_marker(startupsMarker);
+                    
+                
+                }).error(function(data, status, headers, config) {
+                    
+
+                    console.log("There is error.Loklak Server did not respond with geodata.We will try again."+data+"fff"+status);
```
Comment (sudheesh001) [inline app/js/controllers/angel.js]: Codacy Issue found: [Unexpected console statement.](https://www.codacy.com/app/sudheesh1995/loklak_webclient/file/2550181458/issues/source?bid=3288768&fileBranchId=3288808#l203)

File: app/js/controllers/angel.js [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+        else
+        {
+          console.log("This startup has no locations");
+        }
+        
+        startupslocation.push(ele.locations[0].display_name);
+      }
+      
+      })
+      Geocode();
```
Comment (sudheesh001) [inline app/js/controllers/angel.js]: Codacy Issue found: ["Geocode" was used before it was defined](https://www.codacy.com/app/sudheesh1995/loklak_webclient/file/2550181458/issues/source?bid=3288768&fileBranchId=3288808#l133)

File: app/js/controllers/angel.js [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+                'Imagery © <a href="http://mapbox.com">Mapbox</a>',
+            id: 'examples.map-20v6611k'
+        });
+        var basemapObj = {
+            "First Basemap": grayscale
+        };
+      
+      var map2 =L.map('map2',{layers:[grayscale,startupslayer]}).setView([20,0], 2);
+      
+      var code = function () {
```
Comment (sudheesh001) [inline app/js/controllers/angel.js]: Codacy Issue found: [Wrap an immediate function invocation in parentheses.](https://www.codacy.com/app/sudheesh1995/loklak_webclient/file/2550181458/issues/source?bid=3288768&fileBranchId=3288808#l39)

File: app/js/controllers/angel.js [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+        // If second entry with this name
+      } else if (typeof query_string[pair[0]] === "string") {
+          var arr = [ query_string[pair[0]],decodeURIComponent(pair[1]) ];
+           query_string[pair[0]] = arr;
+        // If third or later entry with this name
+      } else {
+        query_string[pair[0]].push(decodeURIComponent(pair[1]));
+        }
+      } 
+    return query_string;
```
Comment (sudheesh001) [inline app/js/controllers/angel.js]: Codacy Issue found: [Identifier 'query_string' is not in camel case.](https://www.codacy.com/app/sudheesh1995/loklak_webclient/file/2550181458/issues/source?bid=3288768&fileBranchId=3288808#l57)

**Comment by leonmak** [comment]
@prasht63 the user accounts will not depend on loklak server, see #594, we will probably use passportjs for auth, the current auth is to use hello.js which does not support angelist

**Comment by mariobehling** [comment]
Closing PR as it is outdated.

**Pull request closed without merging**